### PR TITLE
[TRTLLM-5966][feat] Initial steps towards Helix parallelism support

### PR DESCRIPTION
**Pull request by MatthiasKohl**
# [TRTLLM-5966][feat] Initial steps towards Helix parallelism support

See JIRA ticket https://jirasw.nvidia.com/browse/TRTLLM-5966

## Description

This PR adds initial steps for TRT-LLM to be able to support Helix parallelism:

1. Better support in mapping.py for context parallelism + added unit tests for these cases
2. Added CpType enum in mapping.py to better handle different types of context parallelism: changed this throughout pytorch backend (mainly for star attention)
3. Initial support for Helix in py_executor.py / model_engine.py : including renaming `all_rank_num_tokens` to `all_tp_rank_num_tokens` / `all_cp_rank_num_tokens`. Note: after merging latest main, this is not done throughout the code-base for the AttentionMetadata anymore.
4. Support for Helix in MLA attention
5. Unit test for Helix in MLA attention
6. Test for Helix in disaggregated tests: waived for now as this is WIP
7. Minor changes to attentionOp.hpp/cpp to expose softmax stats for both context and generation

## Test Coverage

- tests/unittest/others/test_mapping.py
- tests/unittest/_torch/modules/test_mla_helix.py

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

* **New Features**
  * Added support for Helix context parallelism with new CP types and configuration options.
  * Introduced distributed collective operations: alltoall and cp_allgather for multi-device communication.
  * Enhanced attention and MLA modules to support additional parallelism dimensions and Helix-specific logic.
  * Added new integration and unit tests for Helix parallelism and multi-GPU distributed scenarios.

* **Improvements**
  * Refined configuration and mapping APIs to use strongly-typed enums for context parallelism types.
  * Consolidated optional tensor arguments in attention kernels for better extensibility and added helix position offsets support.
  * Updated public APIs and documentation to reflect new context parallelism and communication primitives.

* **Bug Fixes**
  * Corrected handling and assignment of softmax statistics pointers in attention operations.

* **Chores**
  * Updated test configurations and skip lists to include new Helix and DeepSeek-V3-Lite scenarios.

### CODE REVIEW

File: tensorrt_llm/_torch/modules/attention.py [hunk 1]
```diff
@@ -802,7 +893,14 @@ def forward_impl(self,
...
             k_pe_gen = k_pe[num_ctx_tokens:, ...]
-            latent_cache_gen = latent_cache[num_ctx_tokens:, ...]
+            # TODO for CP Helix: any rank except the last one / the one generating
+            # the output token should use the latent cache of the "next" logical rank
+            # (first token for each sequence in the KV cache of next rank)
+            # this generally doesn't influence the results much, but it could
+            # for now, we solve it by allowing to pass a custom latent cache for generation
+            latent_cache_gen = latent_cache[
+                num_ctx_tokens:,
+                ...] if latent_cache_gen is None else latent_cache_gen
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply tensorrt_llm/_torch/modules/attention.py:903]: sure give it your best shot

File: tests/unittest/_torch/modules/test_mla_helix.py [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+    print(
+        f"Rank {rank} {world_size}-GPU: max error index: {max_err_idx} "
+        f"(test/ref values: {values_err}), max abs error index: {max_abs_err_idx} "
+        f"(test/ref values: {values_abs}, err: {max_abs_err}), max rel error index: {max_rel_err_idx} "
+        f"(test/ref values: {values_rel}, err: {max_rel_err}), atol: {atol}, rtol: {rtol}"
+    )
+    isclose = err < atol + rtol * ref_abs
+    ratio_mismatch = 0.0
+    if not isclose.all().item():
+        n_mismatch = (isclose == False).sum().item()
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply tests/unittest/_torch/modules/test_mla_helix.py]: addressed in latest commit

**Review by brb-nv — COMMENTED** [review]
Hi Matthias, thank you for the awesome work!

I'm half-way through the review and am worried the current change is too big and is doing too many things. I think it's better to break it down into 4-5 MRs for more focused reviews. My proposal would be to merge in this order:
1) Add alltoall op to TRTLLM
2) Re-group attentionOp args to deal with len<64 requirement
3) Changes to Mapping class
4) Core changes to MLA kernels and attention.py
5) Changes for disagg

Do you think that's a good idea?

File: cpp/tensorrt_llm/thop/attentionOp.cpp [hunk 3]
```diff
@@ -278,17 +282,20 @@ class Runner : public RunnerBase
         common_enqueue_params.context_lengths = context_lengths_ptr;
         common_enqueue_params.host_context_lengths = host_context_lengths.data_ptr<int32_t>();
         common_enqueue_params.workspace = workspace_ptr;
+        if (softmax_stats_tensor.has_value())
+        {
+            TLLM_CHECK_WITH_INFO(softmax_stats_tensor.value().size(-1) == 2,
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply cpp/tensorrt_llm/thop/attentionOp.cpp:287]: it's for both the max and LSE (denominator) of softmax. Since I'm not adding the softmax stats variable to the underlying op in this PR, I'll try to see where it is defined and whether this is already mentioned

File: cpp/tensorrt_llm/common/attentionOp.h [hunk 4]
```diff
@@ -106,6 +106,8 @@ class AttentionOp
         // optional when cross attention
         int32_t const* encoder_input_lengths = nullptr;
         int64_t const* runtime_perf_knobs = nullptr;
+        // optional when compute attention stats (MLA chunked prefill or Helix parallelism)
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply cpp/tensorrt_llm/common/attentionOp.h:109]: no, both chunked prefill and Helix use the exact same output from attention. It is not shared because chunked prefill and Helix will not call attention at the same time.

File: tensorrt_llm/_torch/distributed/communicator.py [hunk 5]
```diff
@@ -100,6 +99,7 @@ class MPIDist(Distributed):
     def __init__(self, mapping: Mapping):
         super().__init__(mapping)
         self.create_tp_comm()
+        self.create_cp_comm()
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply tensorrt_llm/_torch/distributed/communicator.py:102]: the torch backend basically doesn't support any type of CP. The only thing it supports is star attention, which is completely custom (own backend etc.) and AFAIK, star attention cannot be mixed with any of the other parallelisms, or at least only in very specific ways

File: tensorrt_llm/mapping.py [hunk 6]
```diff
@@ -135,34 +147,43 @@ def __init__(
         if moe_cluster_size == -1:
             moe_cluster_size = 1
 
+        cp_type = CpType.ULYSSES if cp_config is None else cp_config.get(
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply tensorrt_llm/mapping.py:150]: This was not added by me, previously Ulysses was the de-facto default (although it was not explicitly mentioned). It is also only for non-torch backend (as mentioned elsewhere, torch backend doesn't support any CP right now).
I have no idea who decided that Ulysses should be the default.

File: tensorrt_llm/mapping.py [hunk 7]
```diff
@@ -372,8 +397,16 @@ def node_rank(self):
...
 
-    def has_cp(self):
-        return self.cp_size > 1
+    def has_cp_ulysses(self):
+        return self.cp_size > 1 and self.cp_config.get(
+            "cp_type", CpType.ULYSSES) == CpType.ULYSSES
+
+    def has_cp_helix(self):
+        return self.cp_size > 1 and self.cp_config.get(
+            "cp_type", CpType.HELIX) == CpType.HELIX
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply tensorrt_llm/mapping.py:406]: Indeed that seems like an issue, I'll fix it.
